pkg-manager/resolve-dependencies/src/resolveDependencies.ts

```diff
@@ -1086,9 +1087,8 @@ async function resolveDependency (
         ? ctx.lockfileDir
         : options.parentPkg.rootDir,
       registry: wantedDependency.alias && pickRegistryForPackage(ctx.registries, wantedDependency.alias, wantedDependency.pref) || ctx.registries.default,
-      // Unfortunately, even when run with --lockfile-only, we need the *real* package.json
-      // so fetching of the tarball cannot be ever avoided. Related issue: https://github.com/pnpm/pnpm/issues/1176
-      skipFetch: false,
+      // we can skip fetching only if we use the NPMJS Files Index, and we don't need the full tarball
+      skipFetch: (ctx.dryRun && !!ctx.useExperimentalNpmjsFilesIndex),
```

Comment:
This should be enough as `useExperimentalNpmjsFilesIndex` is already only `true` when `lockfileOnly` is `true`.

```suggestion
      skipFetch: !!ctx.useExperimentalNpmjsFilesIndex,
```